### PR TITLE
CBL-5752: Lazy Index

### PR DESCRIPTION
**Pull request by bmeike**
At last.  All tests passing.

Maybe not completely caught up with the latest spec.

### CODE REVIEW

File: common/main/java/com/couchbase/lite/internal/core/C4Collection.java [hunk 1]
```diff
@@ -293,7 +293,17 @@ public void createVectorIndex(
...
     @NonNull
     public FLValue getIndexesInfo() throws LiteCoreException {
-        return FLValue.getFLValue(withPeerOrThrow(impl::nGetIndexesInfo));
+        return withPeerOrThrow(peer -> FLValue.getFLValue(impl.nGetIndexesInfo(peer)));
+    }
+
+    @GuardedBy("Database.getDbLock()")
+    @Nullable
+    public C4Index getIndex(@NonNull String name) throws LiteCoreException {
+        // ??? need to check collection for validity?
```
Comment (pasin) [inline common/main/java/com/couchbase/lite/internal/core/C4Collection.java:302]: Should be the same as what the other methods in the Collection class are doing.

File: common/main/cpp/native_c4collection.cc [hunk 2]
```diff
@@ -339,11 +339,48 @@ Java_com_couchbase_lite_internal_core_impl_NativeC4Collection_createVectorIndex(
...
+            return 0;
+
+        // If the index was not found, just return null
+        if ((error.domain != LiteCoreDomain) || (error.code != kC4ErrorMissingIndex))
+            throwError(env, error);
+
+        return 0;
+    }
+
+    c4index_retain(idx);
```
[the remark this answers is not in the document]
Comment (bmeike) [reply common/main/cpp/native_c4collection.cc]: Completely misunderstood the docs.  Fixed.

File: common/main/java/com/couchbase/lite/internal/core/C4Collection.java (hunk 1 above)
Comment (bmeike) [reply common/main/java/com/couchbase/lite/internal/core/C4Collection.java:302]: Yes.  fixed.  Thx!

File: common/main/cpp/native_c4index.cc [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    C4Error error{};
+    C4IndexUpdater *updater = c4index_beginUpdate((C4Index *) handle, (size_t) limit, &error);
+    if (updater == nullptr) {
+        if (error.code != 0)
+            throwError(env, error);
+
+        return 0;
+    }
+
+    c4indexupdater_retain(updater);
```
[the remark this answers is not in the document]
Comment (bmeike) [reply common/main/cpp/native_c4index.cc]: Yeah.  Fixed.  Thx.

File: common/main/cpp/native_c4collection.cc [hunk 4]
```diff
@@ -339,11 +339,48 @@ Java_com_couchbase_lite_internal_core_impl_NativeC4Collection_createVectorIndex(
...
+ */
+JNIEXPORT jlong JNICALL
+Java_com_couchbase_lite_internal_core_impl_NativeC4Collection_getIndex(
+        JNIEnv *env,
+        jclass ignore,
+        jlong coll,
+        jstring jName) {
+    jstringSlice name(env, jName);
+
+    if (!c4coll_isValid((C4Collection *) coll)) {
```
[the remark this answers is not in the document]
Comment (bmeike) [reply common/main/cpp/native_c4collection.cc]: Leaving it to LiteCore....

File: common/main/cpp/native_c4collection.cc [hunk 5]
```diff
@@ -339,11 +339,41 @@ Java_com_couchbase_lite_internal_core_impl_NativeC4Collection_createVectorIndex(
...
+        jclass ignore,
+        jlong coll,
+        jstring jName) {
+    jstringSlice name(env, jName);
+
+    C4Error error{};
+    C4Index *idx = c4coll_getIndex((C4Collection *) coll, name, &error);
+    if (idx == nullptr) {
+        // no error code; no error.  :shrug:
+        if (error.code != 0)
```
Comment (pasin) [inline common/main/cpp/native_c4collection.cc:365]: should this be == 0?

**Review by pasin — CHANGES_REQUESTED** [review]

File: common/main/cpp/native_c4collection.cc (hunk 5 above)
Comment (bmeike) [reply common/main/cpp/native_c4collection.cc]: Boy I'm glad you caught that.  I do that kind of stuff *constantly*

Fixed.

**Review by pasin — APPROVED** [review]

File: common/main/cpp/native_c4index.cc [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+Java_com_couchbase_lite_internal_core_impl_NativeC4IndexUpdater_finish(
+        JNIEnv *env,
+        jclass ignore,
+        jlong handle) {
+#ifdef COUCHBASE_ENTERPRISE
+    C4Error error{};
+    bool ok = c4indexupdater_finish((C4IndexUpdater *) handle, &error);
+    if (!ok && error.code != 0) {
+        throwError(env, error);
+        return;
```
Comment (borrrden) [inline common/main/cpp/native_c4index.cc:174]: I saw a comment in Slack indicating that you should release in an error situation as well.  It looks like you are not doing that here?

Comment (bmeike) [reply common/main/cpp/native_c4index.cc:174]: Yeah... It turns out that test #22 explicitly forbids doing that.  Guess how I found out....

**Pull request merged**